### DISCUSSION TITLE
Why downloading a 177MB file using Webclient took time (~18 sec) vs time taken by Downloader with chunk size = 1 (~107.6 sec)?

### DISCUSSION THREAD

**Discussion by vishal19217** [bezzad/Downloader#178] [Q&A]
Hi @bezzad 
I am trying to use the downloader in my application .
My requirements is to download files of extension(.mp3,.mp4,.zip,.file) over unstable server/ Internet connection.
Files sizes can vary between (1 kb to 1 TB).
1. I am using the downloader nuget package in my project . I started with simple case of  chunk  = 1(over stable/fast internet) but here i found webclient surpassed the downloader in terms of downloading the file? 
2. Can you guide me how to use downloader to met my requirements.

**Comment by bezzad** [comment]
Hi @vishal19217,

Thank you for reaching out and for your interest in using the `Downloader` package in your application.

To better assist you with your requirements, it would be helpful if you could provide a few more details:

* Download Link: Could you share the specific URL you're using for the file download? This will help us identify if there are any server-related factors at play.

* Download Configuration :
What are the specific configurations you're using in the `Downloader` (e.g., chunk size, timeout settings, retries, etc.)?
Are there any specific properties or methods that you're utilizing?
Are you handling any events or callbacks that may impact the download behavior?
Regarding your observation that the `WebClient` seems to outperform the `Downloader` when using a chunk size of 1, it's important to note that the `Downloader` indeed utilizes the `HttpWebRequest` class as the base, which is also used by `WebClient`. 
Here are a few factors that might affect performance:

* Connection Management : `WebClient` may handle connections differently than the `Downloader` due to its internal implementations.
* Overhead : The `Downloader` might introduce some overhead with its additional features for handling resumable downloads, progress tracking, etc.
* File Size and Network Conditions : Given the wide range of file sizes and unstable connections, performance can vary significantly based on those conditions.

To help improve your download experience:

* Chunk Size : Experiment with different chunk sizes. Increasing the chunk size can often lead to better performance on unstable connections since it reduces overhead.
* Retries and Timeouts : Adjusting the retry logic and timeouts may improve the reliability of your downloads, especially when dealing with unstable connections.

Once you provide more details, I’ll be better positioned to give you targeted advice to meet your downloading requirements effectively.

Best regards

**Comment by vishal19217** [comment]
Hi @bezzad Thanks for your response.

Testing url I am using : "http://commondatastorage.googleapis.com/gtv-videos-bucket/sample/TearsOfSteel.mp4". But in my app it would be of some CDN like akamai.
downloader configuration: 

```

Stopwatch stopwatch = Stopwatch.StartNew();
var downloadOpt = new DownloadConfiguration()
            {
                ParallelDownload = true,
                MaxTryAgainOnFailover = 3,
                Timeout = 60000,
                ClearPackageOnCompletionWithFailure = true,
  
            };

            var downloader = new DownloadService(downloadOpt);
            downloader.DownloadStarted += Downloader_DownloadStarted;
            await downloader.DownloadFileTaskAsync(url, filePath);
            stopwatch.Stop();
            Console.WriteLine("The elapsed time for the chunk downloading: "+stopwatch.Elapsed.TotalSeconds);

```
Webclient code:-
```
using(var client = new WebClient() )
            {
                Stopwatch stopwatch = Stopwatch.StartNew();

                await client.DownloadFileTaskAsync(url, fileName);
                stopwatch.Stop();
                Console.WriteLine("The elapsed Time for the files downloaded using Webclient: " + stopwatch.Elapsed.TotalSeconds);
            }
```


2. Also After resolving this, i am thinking to find the efficient no. of chunks i should use since my files can vary from 1kb to 1TB.
I am thinking to use no. of chunks = ContentLength/64kb
and chunk size of 64kb
